homeassistant/components/sensor/recorder.py

```diff
@@ -220,23 +217,22 @@ def _normalize_states(
     fstates = []
 
     for state in entity_history:
-        # Exclude non numerical states from statistics
-        if not _is_number(state.state):
-            continue
+        try:
+            fstate = float(state.state)
+            unit = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
+            # Exclude unsupported units from statistics
+            if unit not in UNIT_CONVERSIONS[device_class]:
+                if WARN_UNSUPPORTED_UNIT not in hass.data:
+                    hass.data[WARN_UNSUPPORTED_UNIT] = set()
+                if entity_id not in hass.data[WARN_UNSUPPORTED_UNIT]:
+                    hass.data[WARN_UNSUPPORTED_UNIT].add(entity_id)
+                    _LOGGER.warning("%s has unknown unit %s", entity_id, unit)
+                continue
 
-        fstate = float(state.state)
-        unit = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
-        # Exclude unsupported units from statistics
-        if unit not in UNIT_CONVERSIONS[device_class]:
-            if WARN_UNSUPPORTED_UNIT not in hass.data:
-                hass.data[WARN_UNSUPPORTED_UNIT] = set()
-            if entity_id not in hass.data[WARN_UNSUPPORTED_UNIT]:
-                hass.data[WARN_UNSUPPORTED_UNIT].add(entity_id)
-                _LOGGER.warning("%s has unknown unit %s", entity_id, unit)
+            fstates.append((UNIT_CONVERSIONS[device_class][unit](fstate), state))
+        except ValueError:
```

Comment:
We should only wrap the code that can raise inside try-except. Use try…except…else to put the code for when nothing got raised.